pycbc/fft/fftw.py

```diff
@@ -406,18 +406,16 @@ def _fftw_setup(fftobj):
     plan_func = _plan_funcs_dict[ (str(fftobj.invec.dtype), str(fftobj.outvec.dtype)) ]
     tmpin = zeros(len(fftobj.invec), dtype = fftobj.invec.dtype)
     tmpout = zeros(len(fftobj.outvec), dtype = fftobj.outvec.dtype)
-    # C2C, forward
-    if fftobj.forward and (fftobj.outvec.dtype in [complex64, complex128]):
-        plan = plan_func(1, n.ctypes.data, fftobj.nbatch,
-                         tmpin.ptr, inembed.ctypes.data, 1, fftobj.idist,
-                         tmpout.ptr, onembed.ctypes.data, 1, fftobj.odist,
-                         FFTW_FORWARD, flags)
-    # C2C, backward
-    elif not fftobj.forward and (fftobj.invec.dtype in [complex64, complex128]):
+    # C2C
+    if fftobj.outvec.kind == 'complex' and fftobj.invec.kind == 'complex':
+        if fftobj.forward:
+            ffd = FFTW_FORWARD
+        else:
+            ffd = FFTW_BACKWARD
         plan = plan_func(1, n.ctypes.data, fftobj.nbatch,
```

Comment:
So I don't think there's anything wrong with this change, but I also don't see that there was anything that should have been wrong with the code you're replacing, unless something was being set wrong about `fftobj.forward` (which could then be causing other problems). I've lost track of which changes were tested incrementally from others---was there a point at which the new lines 409--414 were the only change you made, and it fixed one of the problems you were seeing?  If not, it may be that some of the other changes were actually more responsible for fixing broken behavior; if so, I'm puzzled and mostly want to make sure that there isn't some remaining subtle issue that we are still missing.

When you were seeing the segfault in FFTW, could you tell whether it was in the function that created the plan, or in the function that executed that plan?

Reply:
This was the change that resolved the FFTW segfault. I guess that reverting this and running the tests again would prove that. I think though that refactoring this a little to remove some duplication is hiding the bug-fix in the if statement logic.

The segfault before was in R2C transforms (so C2C was good, as you highlight below). The problem was that if you have R2C you want to use the R2C plan (seen at line 423). *But* the R2C transform actually ended up at line 411 because the logic was not correct (it's a forward transform and the output is complex, so it ends up at the forward C2C plan). The logic needs to look at the type of the input as well as the output, and that's what I added. The rest is just cleanup.

The segfault actually happened when running the FFT, not when planning. The C2R in contrast did run, but gave invalid output.